crates/aggchain-proof-builder/src/lib.rs

```diff
@@ -121,6 +122,162 @@ pub struct AggchainProofBuilderResponse {
+        let key = key_fn(value);
+        if let Some(count) = removal_map.get_mut(&key) {
+            if *count > 0 {
+                *count -= 1;
+                continue;
+            }
+        }
+        result.push(value.clone());
+    }
+
+    Ok(result)
+}
+
+/// Filters, sorts, and maps items from an iterator based on a block number
+/// range.
+///
+/// This function takes an iterator of items, filters them to include only those
+/// whose block number (as determined by `block_number_fn`) falls within the
+/// specified `range`, sorts the filtered items using their `Ord`
+/// implementation, and then maps each item to a new type using the provided
+/// `map_fn`.
+///
+/// # Type Parameters
+/// - `T`: The type of the input items. Must implement `Ord`.
+/// - `F`: The mapping function type. Must be a function or closure that takes
+///   `T` and returns `U`.
+/// - `U`: The type of the output items.
+///
+/// # Arguments
+/// - `items`: An iterator of items to process.
+/// - `range`: The inclusive range of block numbers to filter by.
+/// - `block_number_fn`: A function that extracts the block number from an item.
+/// - `map_fn`: A function that maps each filtered and sorted item to the
+///   desired output type.
+///
+/// # Returns
+/// An iterator over the mapped items, filtered and sorted as described.
+///
+/// # Example
+/// ```no_run
+/// # use aggchain_proof_builder::filter_sort_map;
+/// # struct Item { block_number: u64 }
+/// # impl Item {
+/// #     fn to_output_type(self) -> u64 { self.block_number }
+/// # }
+/// # impl Ord for Item {
+/// #     fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+/// #         self.block_number.cmp(&other.block_number)
+/// #     }
+/// # }
+/// # impl PartialOrd for Item {
+/// #     fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+/// #         Some(self.cmp(other))
+/// #     }
+/// # }
+/// # impl Eq for Item {}
+/// # impl PartialEq for Item {
+/// #     fn eq(&self, other: &Self) -> bool {
+/// #         self.block_number == other.block_number
+/// #     }
+/// # }
+/// let item1 = Item { block_number: 100 };
+/// let item2 = Item { block_number: 150 };
+/// let item3 = Item { block_number: 250 };
+/// let items = vec![item1, item2, item3];
+/// let range = 100..=200;
+/// let result: Vec<_> = filter_sort_map(
+///     items,
+///     &range,
+///     |item| item.block_number,
+///     |item| item.to_output_type(),
+/// )
+/// .collect();
+/// assert_eq!(result, vec![100, 150]);
+/// ```
+pub fn filter_sort_map<T, F, U>(
+    items: impl IntoIterator<Item = T>,
+    range: &std::ops::RangeInclusive<u64>,
+    block_number_fn: fn(&T) -> u64,
+    map_fn: F,
+) -> impl Iterator<Item = U>
+where
+    F: Fn(T) -> U,
+    T: Ord,
+{
+    let mut filtered_items: Vec<_> = items
+        .into_iter()
+        .filter(|item| range.contains(&block_number_fn(item)))
+        .collect();
+    filtered_items.sort();
+    filtered_items.into_iter().map(map_fn)
```

Comment:
Note: itertools has [an extension method for the sort step](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.sorted).